python/pyspark/rdd.py

```diff
@@ -704,7 +704,12 @@ def pipe_objs(out):
                     out.write(s.encode('utf-8'))
                 out.close()
             Thread(target=pipe_objs, args=[pipe.stdin]).start()
-            return (x.rstrip(b'\n').decode('utf-8') for x in iter(pipe.stdout.readline, b''))
+            result = (x.rstrip(b'\n').decode('utf-8') for x in iter(pipe.stdout.readline, b''))
+            pipe.wait()
```

Comment:
Yes I see the problem now. I think it can be overcome by wrapping the pipe.wait and subsequent lines in a second iterator. Something like:
```
from itertools import chain
...
def check_return_code(pipe):
    pipe.wait()
    if pipe.returncode:
        raise Exception("Pipe function `%s' exited "
           "with error code %d" % (command, pipe.returncode))
    else:
        return None
result = (x.rstrip(b'\n').decode('utf-8') for x in chain(iter(pipe.stdout.readline, b''), iter(check_return_code, None)))
```